lib/rspec/core/configuration.rb

```diff
@@ -2027,9 +2031,18 @@ def configure_group_with(group, module_list, application_method)
         end
       end
 
-      def on_existing_matching_groups(meta)
-        world.all_example_groups.each do |group|
-          yield group if meta.empty? || MetadataFilter.apply?(:any?, meta, group.metadata)
+      def on_existing_matching_groups(meta, scope=:ignore)
+        groups = world.example_groups.dup
+
+        until groups.empty?
+          group = groups.shift
+
+          if scope == :example || scope == :each || scope.nil? ||
```

Comment:
This duplicates some very specific hook-related logic:

 - `:example` hooks are always added, even if they don't match the group's metadata: https://github.com/rspec/rspec-core/blob/v3.5.0/lib/rspec/core/hooks.rb#L519
 - `:each` is an alias for `:example`: https://github.com/rspec/rspec-core/blob/v3.5.0/lib/rspec/core/hooks.rb#L478
 - when no scope is passed, `:example` is implied: https://github.com/rspec/rspec-core/blob/v3.5.0/lib/rspec/core/hooks.rb#L586

In the previous implementation this logic were still duplicated, but at least it was still in the `Hooks` module.